File: transport/frame.go
// Package frame implements the DeMoD 17-byte DCF transport frame codec.
//
// Wire layout:
//
//	Byte  Field       Notes
//	────  ─────────   ──────────────────────────────────────────────
//	 0    sync        Fixed 0xD3
//	 1    flags       [7:4] version (4-bit) | [3:0] type (4-bit)
//	 2-3  seq         Big-endian uint16
//	 4-5  src_id      Big-endian uint16
//	 6-7  dst_id      Big-endian uint16  (0xFFFF = broadcast)
//	 8-11 payload     4 raw bytes
//	12-14 timestamp   24-bit big-endian µs offset (wraps ~16.7 s)
//	15-16 crc16       CRC-CCITT(0x1021, init=0xFFFF) over bytes [0..14]
//
// Reference test vector:
//
//	Input:  version=1, DATA, seq=1, src=1, dst=0xFFFF, payload=DEADBEEF, ts=0
//	Output: D3 10 00 01 00 01 FF FF DE AD BE EF 00 00 00 42 DD
//
// The CRC value 0x42DD is pinned across all language implementations.
// If this package's TestCRCPin fails, the codec has diverged from the others.
//
// No external dependencies.
package frame

import "errors"

// FrameSize is the fixed wire size in bytes.
const FrameSize = 17

// SyncByte is the fixed preamble byte.
const SyncByte = 0xD3

// Broadcast is the special dst_id meaning "all nodes".
const Broadcast uint16 = 0xFFFF

// Type represents the 4-bit frame type nibble.
type Type uint8

const (
	TypeData   Type = 0 // application data
	TypeAck    Type = 1 // acknowledgement
	TypeBeacon Type = 2 // clock sync / beacon
	TypeCtrl   Type = 3 // control / fragmentation
)

func (t Type) String() string {
	switch t {
	case TypeData:
		return "DATA"
	case TypeAck:
		return "ACK"
	case TypeBeacon:
		return "BEACON"
	case TypeCtrl:
		return "CTRL"
	default:
		return "UNKNOWN"
	}
}

// Sentinel errors returned by Decode.
var (
	ErrBadLength  = errors.New("frame: buffer must be exactly 17 bytes")
	ErrBadSync    = errors.New("frame: invalid sync byte (expected 0xD3)")
	ErrBadCRC     = errors.New("frame: CRC mismatch")
	ErrUnknownType = errors.New("frame: unknown frame type nibble")
)

// Frame is the decoded representation of a DCF transport frame.
type Frame struct {
	Version     uint8    // 4-bit (0–15)
	Type        Type     // 4-bit
	Seq         uint16   // big-endian rolling counter
	SrcID       uint16
	DstID       uint16
	Payload     [4]byte
	TimestampUs uint32   // lower 24 bits used; upper byte ignored on encode
}

// ── CRC-CCITT (poly 0x1021, init 0xFFFF) ─────────────────────────────────────
// Identical algorithm to C, Haskell, Rust, and Lisp implementations.

func crc16CCITT(data []byte) uint16 {
	crc := uint16(0xFFFF)
	for _, b := range data {
		crc ^= uint16(b) << 8
		for i := 0; i < 8; i++ {
			if crc&0x8000 != 0 {
				crc = (crc << 1) ^ 0x1021
			} else {
				crc <<= 1
			}
		}
	}
	return crc
}

// ── Encode ────────────────────────────────────────────────────────────────────

// Encode serialises f into exactly 17 bytes.
// CRC is computed and appended automatically.
// All multi-byte fields are big-endian.
func (f *Frame) Encode() [FrameSize]byte {
	var w [FrameSize]byte
	w[0] = SyncByte
	w[1] = ((f.Version & 0x0F) << 4) | (uint8(f.Type) & 0x0F)
	w[2] = uint8(f.Seq >> 8)
	w[3] = uint8(f.Seq)
	w[4] = uint8(f.SrcID >> 8)
	w[5] = uint8(f.SrcID)
	w[6] = uint8(f.DstID >> 8)
	w[7] = uint8(f.DstID)
	w[8]  = f.Payload[0]
	w[9]  = f.Payload[1]
	w[10] = f.Payload[2]
	w[11] = f.Payload[3]
	w[12] = uint8((f.TimestampUs >> 16) & 0xFF)
	w[13] = uint8((f.TimestampUs >>  8) & 0xFF)
	w[14] = uint8(f.TimestampUs & 0xFF)
	crc := crc16CCITT(w[:15])
	w[15] = uint8(crc >> 8)
	w[16] = uint8(crc)
	return w
}

// ── Decode ────────────────────────────────────────────────────────────────────

// Decode parses a 17-byte slice into a Frame.
// Returns ErrBadLength if len(b) != 17, ErrBadSync if b[0] != 0xD3,
// ErrBadCRC if the CRC fails, ErrUnknownType for a reserved type nibble.
func Decode(b []byte) (*Frame, error) {
	if len(b) != FrameSize {
		return nil, ErrBadLength
	}
	if b[0] != SyncByte {
		return nil, ErrBadSync
	}
	storedCRC := uint16(b[15])<<8 | uint16(b[16])
	if crc16CCITT(b[:15]) != storedCRC {
		return nil, ErrBadCRC
	}

	typeNib := Type(b[1] & 0x0F)
	if typeNib > TypeCtrl {
		return nil, ErrUnknownType
	}

	f := &Frame{
		Version: (b[1] >> 4) & 0x0F,
		Type:    typeNib,
		Seq:     uint16(b[2])<<8 | uint16(b[3]),
		SrcID:   uint16(b[4])<<8 | uint16(b[5]),
		DstID:   uint16(b[6])<<8 | uint16(b[7]),
		Payload: [4]byte{b[8], b[9], b[10], b[11]},
		TimestampUs: uint32(b[12])<<16 | uint32(b[13])<<8 | uint32(b[14]),
	}
	return f, nil
}

// ── IsValid ───────────────────────────────────────────────────────────────────

// IsValid returns true iff b is a valid 17-byte frame (sync + CRC correct).
// Does not fully decode; cheaper than Decode for hot-path filtering.
func IsValid(b []byte) bool {
	if len(b) != FrameSize || b[0] != SyncByte {
		return false
	}
	stored := uint16(b[15])<<8 | uint16(b[16])
	return crc16CCITT(b[:15]) == stored
}

File: haskell/src/Frame.hs
-- DCF.Transport.Frame — DeMoD 17-byte transport frame
-- DeMoD LLC
--
-- Frame wire layout (17 bytes = 136 bits):
--
--  Byte  Field       Notes
--  ────  ─────────   ──────────────────────────────────────────────────
--   0    sync        Fixed 0xD3
--   1    flags       [7:4] version (4-bit) | [3:0] type (4-bit)
--   2-3  seq         Big-endian Word16
--   4-5  src_id      Big-endian Word16
--   6-7  dst_id      Big-endian Word16  (0xFFFF = broadcast)
--   8-11 payload     4 raw bytes
--  12-14 timestamp   24-bit big-endian µs offset (wraps ~16.7 s)
--  15-16 crc16       CRC-CCITT(0x1021, init=0xFFFF) over bytes [0..14]

{-# LANGUAGE ScopedTypeVariables #-}

module DCF.Transport.Frame
  ( FrameType (..)
  , DeModFrame (..)
  , sealFrame
  , validateFrame
  , encodeFrame
  , decodeFrame
  , frameSyncByte
  , broadcast
  ) where

import Data.Bits   ((.&.), (.|.), shiftL, shiftR, xor)
import Data.Word   (Word8, Word16, Word32)
import Data.ByteString (ByteString)
import qualified Data.ByteString      as BS
import qualified Data.ByteString.Lazy as BL
import Data.Binary.Put (Put, putWord8, putWord16be, runPut)

-- ── Constants ─────────────────────────────────────────────────────────────────

frameSyncByte :: Word8
frameSyncByte = 0xD3

broadcast :: Word16
broadcast = 0xFFFF

-- ── Frame type ────────────────────────────────────────────────────────────────

data FrameType
  = FData    -- 0x0 — application data
  | FAck     -- 0x1 — acknowledgement
  | FBeacon  -- 0x2 — clock sync / beacon
  | FCtrl    -- 0x3 — control / fragmented audio
  deriving (Show, Eq, Ord, Enum, Bounded)

frameTypeToNibble :: FrameType -> Word8
frameTypeToNibble = fromIntegral . fromEnum

nibbleToFrameType :: Word8 -> Maybe FrameType
nibbleToFrameType n
  | n <= 3    = Just (toEnum $ fromIntegral n)
  | otherwise = Nothing

-- ── Frame record ─────────────────────────────────────────────────────────────

data DeModFrame = DeModFrame
  { frameVersion     :: Word8                        -- ^ 4-bit (0-15)
  , frameType        :: FrameType
  , frameSeq         :: Word16                       -- ^ rolling counter
  , frameSrcId       :: Word16                       -- ^ source node
  , frameDstId       :: Word16                       -- ^ dest node (0xFFFF = bcast)
  , framePayload     :: (Word8, Word8, Word8, Word8) -- ^ 4 payload bytes
  , frameTimestampUs :: Word32                       -- ^ lower 24 bits used
  } deriving (Show, Eq)

-- ── CRC-CCITT (poly 0x1021, init 0xFFFF) ─────────────────────────────────────

crc16ccitt :: ByteString -> Word16
crc16ccitt = BS.foldl' step 0xFFFF
  where
    step :: Word16 -> Word8 -> Word16
    step crc b =
      let crc' = crc `xor` (fromIntegral b `shiftL` 8)
      in foldl applyBit crc' ([0..7] :: [Int])
    applyBit :: Word16 -> Int -> Word16
    applyBit c _ =
      if c .&. 0x8000 /= 0
        then (c `shiftL` 1) `xor` 0x1021
        else  c `shiftL` 1

-- ── Serialisation ─────────────────────────────────────────────────────────────

putFrameBody :: DeModFrame -> Put
putFrameBody f = do
  putWord8 frameSyncByte
  let flags = ((frameVersion f .&. 0x0F) `shiftL` 4)
            .|. (frameTypeToNibble (frameType f) .&. 0x0F)
  putWord8 flags
  putWord16be (frameSeq   f)
  putWord16be (frameSrcId f)
  putWord16be (frameDstId f)
  let (p0, p1, p2, p3) = framePayload f
  putWord8 p0; putWord8 p1; putWord8 p2; putWord8 p3
  let ts = frameTimestampUs f
  putWord8 (fromIntegral $ (ts `shiftR` 16) .&. 0xFF)
  putWord8 (fromIntegral $ (ts `shiftR`  8) .&. 0xFF)
  putWord8 (fromIntegral $  ts              .&. 0xFF)

-- | Produce the sealed 17-byte frame ByteString.
encodeFrame :: DeModFrame -> ByteString
encodeFrame f =
  let body  = BL.toStrict . runPut $ putFrameBody f
      crc   = crc16ccitt body
      crcHi = fromIntegral (crc `shiftR` 8) :: Word8
      crcLo = fromIntegral (crc .&. 0xFF)   :: Word8
  in body <> BS.pack [crcHi, crcLo]

-- | Validate and decode a 17-byte ByteString.
--   Returns Nothing if the sync byte is wrong or CRC fails.
decodeFrame :: ByteString -> Maybe DeModFrame
decodeFrame bs
  | BS.length bs /= 17             = Nothing
  | BS.index bs 0 /= frameSyncByte = Nothing
  | otherwise =
      let body      = BS.take 15 bs
          crcStored = (fromIntegral (BS.index bs 15) `shiftL` 8)
                    .|. fromIntegral (BS.index bs 16) :: Word16
          crcCalc   = crc16ccitt body
      in if crcCalc /= crcStored
           then Nothing
           else
             let flags   = BS.index bs 1
                 ver     = (flags `shiftR` 4) .&. 0x0F
                 typeNib = flags .&. 0x0F
                 ftype   = case nibbleToFrameType typeNib of
                              Just t  -> t
                              Nothing -> FData
                 seqNum  = (fromIntegral (BS.index bs 2) `shiftL` 8)
                         .|. fromIntegral (BS.index bs 3)
                 srcId   = (fromIntegral (BS.index bs 4) `shiftL` 8)
                         .|. fromIntegral (BS.index bs 5)
                 dstId   = (fromIntegral (BS.index bs 6) `shiftL` 8)
                         .|. fromIntegral (BS.index bs 7)
                 payload = ( BS.index bs  8, BS.index bs  9
                           , BS.index bs 10, BS.index bs 11 )
                 ts      = (fromIntegral (BS.index bs 12) `shiftL` 16)
                         .|. (fromIntegral (BS.index bs 13) `shiftL` 8)
                         .|.  fromIntegral (BS.index bs 14)
             in Just DeModFrame
                  { frameVersion     = ver
                  , frameType        = ftype
                  , frameSeq         = seqNum
                  , frameSrcId       = srcId
                  , frameDstId       = dstId
                  , framePayload     = payload
                  , frameTimestampUs = ts
                  }

-- ── Convenience aliases ───────────────────────────────────────────────────────

sealFrame :: DeModFrame -> ByteString
sealFrame = encodeFrame

validateFrame :: ByteString -> Maybe DeModFrame
validateFrame = decodeFrame
